aclk/agent_cloud_link.c

```diff
@@ -893,6 +946,402 @@ static void aclk_main_cleanup(void *ptr)
+    if (payload != NULL)
+        buffer_strcat(b, payload);
+    debug(D_ACLK, "Sending HTTPS req (%zu bytes): '%s'", b->len, buffer_tostring(b));
+    int sock = connect_to_this_ip46(IPPROTO_TCP, SOCK_STREAM, host, 0, port, &timeout);
+
+    if (unlikely(sock == -1)) {
+        error("Handshake failed");
+        return 1;
+    }
+
+    SSL_CTX *ctx = security_initialize_openssl_client();
+    // Certificate chain: not updating the stores - do we need private CA roots?
+    // Calls to SSL_CTX_load_verify_locations would go here.
+    SSL *ssl = SSL_new(ctx);
+    SSL_set_fd(ssl, sock);
+    int err = SSL_connect(ssl);
+    if (err!=1) {
+        error("SSL_connect() failed with err=%d", err);
+        SSL_free(ssl);
+        close(sock);
+        SSL_CTX_free(ctx);
+        return 1;
+    }
+    err = SSL_write(ssl, b->buffer, b->len);
+    if (err <= 0)
+    {
+        error("SSL_write() failed with err=%d", err);
+        SSL_free(ssl);
+        close(sock);
+        SSL_CTX_free(ctx);
+        return 1;
+    }
+    buffer_flush(b);
+    int bytes_read = SSL_read(ssl, b->buffer, b->size);
+    debug(D_ACLK, "Received %d bytes in response", bytes_read);
+    b->len = bytes_read;
+    SSL_shutdown(ssl);
+    SSL_free(ssl);
+    close(sock);
+        SSL_CTX_free(ctx);
+    return 0;
+}
+
+struct dictionary_singleton {
+    char *key;
+    char *result;
+};
+
+int json_extract_singleton(JSON_ENTRY *e)
+{
+    struct dictionary_singleton *data = e->callback_data;
+
+    switch (e->type) {
+        case JSON_OBJECT:
+        case JSON_ARRAY:
+            break;
+        case JSON_STRING:
+            if (!strcmp(e->name, data->key)) {
+                data->result = strdupz(e->data.string);
+                break;
+            }
+            break;
+        case JSON_NUMBER:
+        case JSON_BOOLEAN:
+        case JSON_NULL:
+            break;
+    }
+    return 0;
+}
+
+// Base-64 decoder.
+// Note: This is non-validating, invalid input will be decoded without an error.
+//       Challenges are packed into json strings so we don't skip newlines.
+//       Size errors (i.e. invalid input size or insufficient output space) are caught.
+size_t base64_decode(unsigned char *input, size_t input_size, unsigned char *output, size_t output_size)
+{
+    static char lookup[256];
+    static int first_time=1;
+    if (first_time)
+    {
+        first_time = 0;
+        for(int i=0; i<256; i++)
+            lookup[i] = -1;
+        for(int i='A'; i<='Z'; i++)
+            lookup[i] = i-'A';
+        for(int i='a'; i<='z'; i++)
+            lookup[i] = i-'a' + 26;
+        for(int i='0'; i<='9'; i++)
+            lookup[i] = i-'0' + 52;
+        lookup['+'] = 62;
+        lookup['/'] = 63;
+    }
+    if ((input_size & 3) != 0)
+    {
+        error("Can't decode base-64 input length %zu", input_size);
+        return 0;
+    }
+    size_t unpadded_size = (input_size/4) * 3;
+    if ( (input_size/4) * 3 > output_size )
```

Comment:
You can use `unpadded_size` here.

Reply:
Ah yes, that was why I calculated it :) Fixed.